src/dataset.jl

```diff
@@ -1013,9 +1013,23 @@ function buildoverviews!(
     return dataset
 end
 
-function destroy(dataset::AbstractDataset)::Nothing
-    GDAL.gdalclose(dataset)
+# TODO: Wrap `GDAL.CPLErr`
+function close(dataset::AbstractDataset)::GDAL.CPLErr
+    dataset.ptr == C_NULL && return GDAL.CE_Failure
+    if !isnothing(dataset.children)
+        for child in dataset.children
+            value = child.value
+            !isnothing(value) && destroy(value)
+        end
+        Base.empty!(dataset.children)
+    end
+    err = GDAL.gdalclose(dataset)
     dataset.ptr = C_NULL
+    return err
+end
+
+function destroy(dataset::AbstractDataset)::Nothing
+    close(dataset)
```

Comment:
So, I would prefer to not change the behavior of `destroy()` (since it's out for quite a while by now), and instead for it to be a new function being introduced e.g. `force_close_mdarray_dataset!(...)` (inside the `src/mdarray/*` subdirectory), and to have it document that it will close all of its children (possibly recursively too)?

We can keep it scoped to `mdarray` is until we have verifiable claims that it's working robustly for the other types of GDAL datasets (FeatureCollections, etc). Once there is a verifiably general implementation that's working and tested for all GDAL dataset types, we can give it a more general name.